### DISCUSSION TITLE
Creating new users needlessly

### DISCUSSION THREAD

**Discussion by tommhuth** [nextauthjs/next-auth#1626] [Help]
When using `next-auth` (locally, not sure if that matters?) I've noticed that it will often, but not always, create a new user instead of reusing my previous one when logging back in. This happens when using the same auth provider with the same credentials.  This seems to happen mostly with GitHub though I'm sure I've experienced it with other providers as well. 

What are the variables involved when `next-auth` decides if a user already exists? Is there anything special going on when running locally? Can `next-auth` not guarantee the same user when logging back in with the same credentials and provider?

**Comment by balazsorban44** [comment]
The behavior you are describing is very odd. I assume that you are using a database, not a session cookie, right? If I'm not mistaken, we use the user ID to identify the user, in github's case, the default is: https://github.com/nextauthjs/next-auth/blob/073da60c3d6c06cfa0ee0918cba0d65b55faf898/src/providers/github.js#L13

sharing some code would help to figure out your problem though.

**Comment by tommhuth** [comment]
Thanks for the speedy feedback:) I was not sure where to begin when debugging this, hence the vagueness. I would say i'm running a pretty basic setup, using `mongodb`:

```js
// api/auth/[nextauth].js
export default NextAuth({ 
    providers: [
        Providers.Facebook({
            clientId: process.env.FACEBOOK_ID,
            clientSecret: process.env.FACEBOOK_SECRET
        }),
        Providers.Google({
            clientId: process.env.GOOGLE_ID,
            clientSecret: process.env.GOOGLE_SECRET
        }),
        Providers.GitHub({
            clientId: process.env.GITHUB_ID,
            clientSecret: process.env.GITHUB_SECRET,
        }),
    ],
    adapter: Adapters.TypeORM.Adapter( 
        process.env.DATABASE_URL, 
        {
            models
        }
    ), 
    database: process.env.DATABASE_URL, 
    secret: process.env.SECRET, 
    session: {    }, 
    jwt: { 
        secret: 'secret', 
    }, 
    pages: {
        signIn: '/login',  // Displays signin buttons 
    }, 
    callbacks: {  
        async session(session, user) {
            return {
                ...session,
                user: {
                    ...session.user,
                    ...user
                }
            }
        }, 
    }, 
    debug: false,
})
```

where `models` refers to this:

```js 


import Adapters from "next-auth/adapters"
 
class User extends Adapters.TypeORM.Models.User.model { 
    constructor(name, email, image, emailVerified) {
        super(name, email, image, emailVerified)
    }
}

const UserSchema = {
    name: "User",
    target: User,
    columns: {
        ...Adapters.TypeORM.Models.User.schema.columns, 
        isOnboarded: {
            type: "boolean",
            nullable: true,
        },
        email: {
            type: "string",
            nullable: true
        },
        username: {
            type: "string",
            nullable: false
        }
    },
}

export default {
    User: {
        model: User,
        schema: UserSchema
    }
}
```

**Reply by DiegoMcDipster** [reply]
I had a similar issue with the email provider.  It was because the user.name field was null.  And I know that with the github provider the email field is null (at least on postgres using prisma adapter).  So, check the fields on your DB for the fields that are null and make sure in your code you initialise them to emptry string.

That solved it for me.  hope it helps.

**Reply by znwhite** [reply]
@DiegoMcDipster How would you recommend initializing the value? Using the `signIn` callback?

**Reply by DiegoMcDipster** [reply]
Yeah... i should've given a bit more information... what can i say... it was late...

So, a bit more detail:
1. Email Provider creates a user account on my Postgres DB using the prisma adapter.  the email address is unique. there is no user.name, it's null.
2. So, when the user logs in they don't have a user.name just a user.email.
3. I created a Formik form to allow the user to enter their name.
4. The formik form has an initialvalues object.
5. If i let the initial values user.name = null the PUT request would create an extra user row on the user table in postgres
6. if i set the initial value of user.name ="" the PUT request would do its job and the user.name would update the one and only user record.

**Reply by tommhuth** [reply]
I'm not sure how this is useful in my case since I'm not using the `EmailProvider`?

**Reply by DiegoMcDipster** [reply]
What i'm saying is... the PUT request of my adapter caused a problem that gave me the same results. I also know that the github provider doesn't fill the name field.  So test it a bit to see if your adapter is doing the same.  It could be a dead end.  Or it could point you in the right direction.

Happy hunting!
